A tool for converting miqldap auth to external auth with sssd

jvlcek opened the pull request:
MiQ is delivered with the built in LDAP client, MiqLdap. Maintaining an MiQ proprietary LDAP
client is unnecessary as there are multiple mechanisms built into the operating system that provide an interface to an LDAP directory.

This PR provides a tool that will convert an MiQ appliances authentication configuration from
using the MiQ proprietary LDAP client to using SSSD packaged into the operating system.

The user records created in the DB by the MiqLdap client were assigned a userid of either
a _**Distinguished Name**_ when backed to an LDAP directory or a _**User Principal Name**_
when backed to an Active Directory, configured as an LDAP directory. This tool also, optionally,
changes all userids to  _**User Principal Name**_ format.

Related PR, [15535](https://github.com/ManageIQ/manageiq/pull/15535) updates external authentication to always store the user record in the DB with a userid in 
_**User Principal Name**_ format. This tool can be used to convert MiQ authentication
from mode: LDAP(S) to SSSD however until PR [15535](https://github.com/ManageIQ/manageiq/pull/15535) is merged duplicate user records will be created.

Steps for Testing/QA [Optional]
-------------------------------

Configure MiQ for both authentication modes of LDAP and LDAPS against both an
LDAP directory and an Active Directory.

Login with a valid user confirm the user record is created in the DB.

Run this tool to convert the authentication mode to external auth with SSSD

Login with the same valid user from step 2 and confirm there is only 1 user record with
a userid in _**User Principal Name**_ format.

jvlcek commented:
@miq-bot remove_label wip

jvlcek commented:
PR [134](https://github.com/ManageIQ/manageiq-appliance/pull/134) is also needed for PR.

jvlcek commented:
@abellotti, @gtanzillo and @jrafanie I remove WIP.

jrafanie commented on tools/miqldap_to_sssd/cli.rb:
```diff
@@ -0,0 +1,57 @@
+            "Do the MiqLdap to SSSD conversion but skip the normalizing of the userids",
+            :short   => "s",
+            :default => false,
+            :type    => :flag
```
In a followup, let's create a separate executable in the bin directory that will invoke this script and allow a user to just run "only_change_userids"

jvlcek replied:
@jrafanie Thank you! I've updated PR https://github.com/ManageIQ/manageiq-appliance/pull/134 to include the requested executable script.

abellotti commented on tools/miqldap_to_sssd/configure_selinux.rb:
```diff
@@ -0,0 +1,55 @@
+  class ConfigureSELinux
+    def configure
+      LOGGER.debug("Invokded #{self.class}\##{__method__}")
+      enable_non_standard_ldap_port("10636")
```
should only enable an ldap port if not 389 and not 636

jvlcek commented:
@gtanzillo As discussed in the code walkthrough, I've update PR [134](https://github.com/ManageIQ/manageiq-appliance/pull/134) to include a second script, _normalize_userid_to_upn_, that will allow the userid field of the user records  to be normalized to UPN format independently of the conversion from miqldap to sssd.

I've also updated the configuration of selinux to be done only for non-standard ldap ports.

I've performed the following testing for upstream and downstream:

-   Run normalize_userid_to_upn before converting miqldap auth to sssd auth
-   Run normalize_userid_to_upn after converting miqldap auth to sssd auth when the userid's were not updated
-   Run normalize_userid_to_upn after converting miqldap auth to sssd auth when the userid's were updated

I've squashed all of my commits down to 1.

With these changes I believe this PR and PR PR [134](https://github.com/ManageIQ/manageiq-appliance/pull/134) are ready to be merged.

Thank you @abellotti, @gtanzillo and @jrafanie for all the help!

abellotti commented:
Great work @jvlcek, I _KNOW_ this was tons of research, investigation, coding and testing that you did for all the different configurations.  Thank you for this tool !!
